File: typescript/packages/mechanisms/evm/test/unit/auth-capture/nonce.test.ts
import { describe, it, expect, vi } from "vitest";
import { keccak256, zeroAddress } from "viem";
import {
  AUTH_CAPTURE_DEPLOYMENT_V1_0,
  AUTH_CAPTURE_DEPLOYMENT_V1_1,
  AUTH_CAPTURE_ESCROW_V1_0_ADDRESS,
  AUTH_CAPTURE_ESCROW_V1_1_ADDRESS,
  CAPTURE_TYPES_V1_0,
  CAPTURE_TYPES_V1_1,
  CHARGE_TYPES_V1_0,
  CHARGE_TYPES_V1_1,
  SALT_BINDING_TYPEHASH,
  captureTypesForDeployment,
  chargeTypesForDeployment,
  feeAmountFromBps,
  resolveAuthCaptureDeployment,
} from "../../../src/auth-capture/constants";
import {
  computePayerAgnosticPaymentInfoHash,
  computePaymentInfoHash,
  deriveBoundSalt,
  extraAddress,
  generateSalt,
  isNonZeroAddress,
  isSaltBindingOn,
  normalizeBytes32,
  signPermit2,
  verifyERC3009Signature,
  verifyPermit2Signature,
} from "../../../src/auth-capture/nonce";
import type { PaymentInfoStruct } from "../../../src/auth-capture/types";

describe("nonce utilities", () => {
  describe("computePayerAgnosticPaymentInfoHash", () => {
    const mockPaymentInfo: PaymentInfoStruct = {
      operator: "0x1111111111111111111111111111111111111111",
      payer: "0xPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP".toLowerCase() as `0x${string}`,
      receiver: "0x2222222222222222222222222222222222222222",
      token: "0x3333333333333333333333333333333333333333",
      maxAmount: "1000000",
      preApprovalExpiry: 281474976710655,
      authorizationExpiry: 281474976710655,
      refundExpiry: 281474976710655,
      minFeeBps: 0,
      maxFeeBps: 100,
      feeReceiver: "0x4444444444444444444444444444444444444444",
      salt: "0x0000000000000000000000000000000000000000000000000000000000000001",
    };

    it("should produce a 32-byte hex string", () => {
      const nonce = computePayerAgnosticPaymentInfoHash(84532, mockPaymentInfo);
      expect(nonce).toMatch(/^0x[a-fA-F0-9]{64}$/);
    });

    it("should produce fixed hashes for the default and v1.0 escrow domains", () => {
      expect(computePayerAgnosticPaymentInfoHash(84532, mockPaymentInfo)).toBe(
        "0x695990db5fd8f3f541f505b117da619dace4d28a039e35646ce7b660e699ff4b",
      );
      expect(computePayerAgnosticPaymentInfoHash(8453, mockPaymentInfo)).toBe(
        "0x37588eff80093203c128c6622608c6e972e41c2fd79da91ef5294a26e647e4e2",
      );
      expect(
        computePayerAgnosticPaymentInfoHash(
          84532,
          mockPaymentInfo,
          AUTH_CAPTURE_ESCROW_V1_0_ADDRESS,
        ),
      ).toBe("0x19de8ffcb747e5caadb3dda7435cf54992e87cdf0c90e5315ffa129dbb22461e");
      expect(
        computePayerAgnosticPaymentInfoHash(
          8453,
          mockPaymentInfo,
          AUTH_CAPTURE_ESCROW_V1_0_ADDRESS,
        ),
      ).toBe("0x198bbfeaab2f8e36302c662ae41bceaef47a1eb4bf2549cb87aa8daa7f7bb43a");
    });

    it("should produce deterministic results for same inputs", () => {
      const nonce1 = computePayerAgnosticPaymentInfoHash(84532, mockPaymentInfo);
      const nonce2 = computePayerAgnosticPaymentInfoHash(84532, mockPaymentInfo);
      expect(nonce1).toBe(nonce2);
    });

    it("should produce different results for different chainIds", () => {
      const nonce1 = computePayerAgnosticPaymentInfoHash(84532, mockPaymentInfo);
      const nonce2 = computePayerAgnosticPaymentInfoHash(8453, mockPaymentInfo);
      expect(nonce1).not.toBe(nonce2);
    });

    it("should produce different results for different payment info", () => {
      const nonce1 = computePayerAgnosticPaymentInfoHash(84532, mockPaymentInfo);
      const nonce2 = computePayerAgnosticPaymentInfoHash(84532, {
        ...mockPaymentInfo,
        maxAmount: "2000000",
      });
      expect(nonce1).not.toBe(nonce2);
    });

    it("should produce different results for different salts (freshness check)", () => {
      const nonce1 = computePayerAgnosticPaymentInfoHash(84532, mockPaymentInfo);
      const nonce2 = computePayerAgnosticPaymentInfoHash(84532, {
        ...mockPaymentInfo,
        salt: "0x0000000000000000000000000000000000000000000000000000000000000002",
      });
      expect(nonce1).not.toBe(nonce2);
    });

    it("computePaymentInfoHash binds the real payer and therefore differs from the nonce", () => {
      const withPayer: PaymentInfoStruct = {
        ...mockPaymentInfo,
        payer: "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
      };
      const nonce = computePayerAgnosticPaymentInfoHash(84532, withPayer);
      const paymentHash = computePaymentInfoHash(84532, withPayer);
      expect(paymentHash).toMatch(/^0x[a-fA-F0-9]{64}$/);
      expect(paymentHash).not.toBe(nonce);
      expect(computePaymentInfoHash(84532, { ...withPayer, payer: zeroAddress })).toBe(nonce);
    });

    it("should be payer-agnostic — different payers produce identical nonces", () => {
      const nonceA = computePayerAgnosticPaymentInfoHash(84532, {
        ...mockPaymentInfo,
        payer: "0xAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA".toLowerCase() as `0x${string}`,
      });
      const nonceB = computePayerAgnosticPaymentInfoHash(84532, {
        ...mockPaymentInfo,
        payer: "0xBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB".toLowerCase() as `0x${string}`,
      });
      const nonceZero = computePayerAgnosticPaymentInfoHash(84532, {
        ...mockPaymentInfo,
        payer: zeroAddress,
      });
      expect(nonceA).toBe(nonceB);
      expect(nonceA).toBe(nonceZero);
    });
  });

  describe("generateSalt", () => {
    it("should produce a 32-byte hex string", () => {
      const salt = generateSalt();
      expect(salt).toMatch(/^0x[a-fA-F0-9]{64}$/);
    });

    it("should produce unique values on each call", () => {
      const salt1 = generateSalt();
      const salt2 = generateSalt();
      const salt3 = generateSalt();
      expect(salt1).not.toBe(salt2);
      expect(salt2).not.toBe(salt3);
      expect(salt1).not.toBe(salt3);
    });

    it("should produce valid hex characters only", () => {
      const salt = generateSalt();
      const hexPart = salt.slice(2);
      expect(hexPart).toMatch(/^[0-9a-f]+$/);
    });
  });

  describe("salt binding", () => {
    const authorizer = "0x1111111111111111111111111111111111111111" as `0x${string}`;
    const policy = "0x0000000000000000000000000000000000000000" as `0x${string}`;
    const nonce =
      "0x0000000000000000000000000000000000000000000000000000000000000abc" as `0x${string}`;

    it("is off when receiverAuthorizer and policy are absent or zero", () => {
      expect(isSaltBindingOn({})).toBe(false);
      expect(isSaltBindingOn({ receiverAuthorizer: policy, policy })).toBe(false);
    });

    it("is on when receiverAuthorizer is non-zero", () => {
      expect(isSaltBindingOn({ receiverAuthorizer: authorizer })).toBe(true);
    });

    it("derives a deterministic 32-byte salt from the typehash, addresses, and nonce", () => {
      const a = deriveBoundSalt(authorizer, policy, nonce);
      const b = deriveBoundSalt(authorizer, policy, nonce);
      expect(a).toBe(b);
      expect(a).toMatch(/^0x[a-fA-F0-9]{64}$/);
      expect(a).toBe("0xd0967e09b6c8fccf96277d95a03e98583e8605ab10858b1349aa50ea6d78132c");
    });

    it("should hash the salt-binding and PaymentInfo type strings to stable values", () => {
      expect(SALT_BINDING_TYPEHASH).toBe(
        "0x8a2a7e41a0bda000ded071ff38b79401d2603e1826516ff2635b11fe9e30877f",
      );
      expect(
        keccak256(
          new TextEncoder().encode(
            "PaymentInfo(address operator,address payer,address receiver,address token,uint120 maxAmount,uint48 preApprovalExpiry,uint48 authorizationExpiry,uint48 refundExpiry,uint16 minFeeBps,uint16 maxFeeBps,address feeReceiver,uint256 salt)",
          ),
        ),
      ).toBe("0xae68ac7ce30c86ece8196b61a7c486d8f0061f575037fbd34e7fe4e2820c6591");
    });

    it("changes when any bound input changes", () => {
      const base = deriveBoundSalt(authorizer, policy, nonce);
      expect(deriveBoundSalt("0x2222222222222222222222222222222222222222", policy, nonce)).not.toBe(
        base,
      );
      expect(
        deriveBoundSalt(
          authorizer,
          policy,
          "0x0000000000000000000000000000000000000000000000000000000000000abd",
        ),
      ).not.toBe(base);
    });
  });

  describe("normalizeBytes32", () => {
    it("zero-pads a short hex integer and lowercases it", () => {
      expect(normalizeBytes32("0xAbC")).toBe(
        "0x0000000000000000000000000000000000000000000000000000000000000abc",
      );
    });

    it("accepts a value without a 0x prefix and a 0X prefix", () => {
      expect(normalizeBytes32("1")).toBe(
        "0x0000000000000000000000000000000000000000000000000000000000000001",
      );
      expect(normalizeBytes32("0Xff")).toBe(
        "0x00000000000000000000000000000000000000000000000000000000000000ff",
      );
    });

    it("rejects empty, non-hex, and oversized values", () => {
      expect(() => normalizeBytes32("0x")).toThrow("Invalid bytes32");
      expect(() => normalizeBytes32("0xzz")).toThrow("Invalid bytes32");
      expect(() => normalizeBytes32(`0x${"aa".repeat(33)}`)).toThrow("Invalid bytes32");
    });
  });

  describe("extraAddress / isNonZeroAddress", () => {
    it("treats absent or invalid addresses as the zero address", () => {
      expect(extraAddress(undefined)).toBe(zeroAddress);
      expect(extraAddress("")).toBe(zeroAddress);
      expect(extraAddress("not-an-address")).toBe(zeroAddress);
      expect(isNonZeroAddress(undefined)).toBe(false);
      expect(isNonZeroAddress("0x0000000000000000000000000000000000000000")).toBe(false);
    });

    it("checksums a valid address and treats it as non-zero", () => {
      expect(extraAddress("0x1111111111111111111111111111111111111111")).toBe(
        "0x1111111111111111111111111111111111111111",
      );
      expect(isNonZeroAddress("0x1111111111111111111111111111111111111111")).toBe(true);
    });
  });

  describe("signature helpers", () => {
    it("verifyPermit2Signature forwards Permit2 typed data to the facilitator signer", async () => {
      const verifyTypedData = vi.fn().mockResolvedValue(true);
      const getCode = vi.fn().mockResolvedValue("0x");
      const signer = {
        getAddresses: () => [],
        readContract: vi.fn(),
        verifyTypedData,
        writeContract: vi.fn(),
        sendTransaction: vi.fn(),
        waitForTransactionReceipt: vi.fn(),
        getCode,
      };
      const permit = {
        from: "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" as `0x${string}`,
        permitted: {
          token: "0x036CbD53842c5426634e7929541eC2318f3dCF7e" as `0x${string}`,
          amount: "1000",
        },
        spender: "0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" as `0x${string}`,
        nonce: "1",
        deadline: "9999999999",
      };

      await expect(verifyPermit2Signature(signer, permit, "0xdead", 84532)).resolves.toBe(false);

      const signed = await signPermit2(
        {
          address: permit.from,
          signTypedData: vi.fn().mockResolvedValue("0xsigned" as `0x${string}`),
        },
        permit,
        84532,
      );
      expect(signed).toBe("0xsigned");
    });

    it("verifyERC3009Signature rejects an invalid signature", async () => {
      const signer = {
        getAddresses: () => [],
        readContract: vi.fn(),
        verifyTypedData: vi.fn().mockResolvedValue(false),
        writeContract: vi.fn(),
        sendTransaction: vi.fn(),
        waitForTransactionReceipt: vi.fn(),
        getCode: vi.fn().mockResolvedValue("0x"),
      };
      const ok = await verifyERC3009Signature(
        signer,
        {
          from: "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
          to: "0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
          value: "1",
          validAfter: "0",
          validBefore: "9",
          nonce: "0x1111111111111111111111111111111111111111111111111111111111111111",
        },
        "0x00",
        { name: "USDC", version: "2", chainId: 84532 },
        "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
      );
      expect(ok).toBe(false);
    });
  });
});

describe("auth-capture deployment constants", () => {
  it("resolves v1.1 by default and v1.0 from the known escrow address", () => {
    expect(resolveAuthCaptureDeployment()).toBe(AUTH_CAPTURE_DEPLOYMENT_V1_1);
    expect(resolveAuthCaptureDeployment("")).toBe(AUTH_CAPTURE_DEPLOYMENT_V1_1);
    expect(resolveAuthCaptureDeployment(AUTH_CAPTURE_ESCROW_V1_1_ADDRESS)).toBe(
      AUTH_CAPTURE_DEPLOYMENT_V1_1,
    );
    expect(resolveAuthCaptureDeployment(AUTH_CAPTURE_ESCROW_V1_1_ADDRESS.toLowerCase())).toBe(
      AUTH_CAPTURE_DEPLOYMENT_V1_1,
    );
    expect(resolveAuthCaptureDeployment(AUTH_CAPTURE_ESCROW_V1_0_ADDRESS)).toBe(
      AUTH_CAPTURE_DEPLOYMENT_V1_0,
    );
    expect(resolveAuthCaptureDeployment("not-an-address")).toBeUndefined();
    expect(
      resolveAuthCaptureDeployment("0x0000000000000000000000000000000000000001"),
    ).toBeUndefined();
  });

  it("selects v1.0 vs v1.1 operator typed-data fields and computes escrow fee amounts", () => {
    expect(chargeTypesForDeployment(AUTH_CAPTURE_DEPLOYMENT_V1_0)).toBe(CHARGE_TYPES_V1_0);
    expect(chargeTypesForDeployment(AUTH_CAPTURE_DEPLOYMENT_V1_1)).toBe(CHARGE_TYPES_V1_1);
    expect(captureTypesForDeployment(AUTH_CAPTURE_DEPLOYMENT_V1_0)).toBe(CAPTURE_TYPES_V1_0);
    expect(captureTypesForDeployment(AUTH_CAPTURE_DEPLOYMENT_V1_1)).toBe(CAPTURE_TYPES_V1_1);
    expect(feeAmountFromBps(1_000_000n, 250)).toBe(25_000n);
    expect(feeAmountFromBps(1n, 1)).toBe(0n);
  });
});

File: go/mechanisms/evm/auth-capture/nonce_test.go
package authcapture

import (
	"math/big"
	"testing"
)

func TestResolveAuthCaptureDeployment(t *testing.T) {
	v11 := ResolveAuthCaptureDeployment("")
	if v11 == nil || v11.Version != AuthCaptureDeploymentV1_1 {
		t.Fatalf("default deployment = %+v", v11)
	}
	if v11.Escrow != AuthCaptureEscrowV1_1Address {
		t.Fatalf("default escrow = %q", v11.Escrow)
	}

	pinnedV11 := ResolveAuthCaptureDeployment(AuthCaptureEscrowV1_1Address)
	if pinnedV11 == nil || pinnedV11.Version != AuthCaptureDeploymentV1_1 {
		t.Fatalf("v1.1 pin = %+v", pinnedV11)
	}

	v10 := ResolveAuthCaptureDeployment(AuthCaptureEscrowV1_0Address)
	if v10 == nil || v10.Version != AuthCaptureDeploymentV1_0 {
		t.Fatalf("v1.0 pin = %+v", v10)
	}
	if v10.EIP3009Collector != EIP3009TokenCollectorV1_0Address {
		t.Fatalf("v1.0 collector = %q", v10.EIP3009Collector)
	}

	if ResolveAuthCaptureDeployment("0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef") != nil {
		t.Fatal("expected unknown escrow to be rejected")
	}
	if ResolveAuthCaptureDeployment("not-an-address") != nil {
		t.Fatal("expected invalid address to be rejected")
	}
}

func mockPaymentInfo() PaymentInfoStruct {
	return PaymentInfoStruct{
		Operator:            "0x1111111111111111111111111111111111111111",
		Payer:               "0xpppppppppppppppppppppppppppppppppppppppp",
		Receiver:            "0x2222222222222222222222222222222222222222",
		Token:               "0x3333333333333333333333333333333333333333",
		MaxAmount:           "1000000",
		PreApprovalExpiry:   281474976710655,
		AuthorizationExpiry: 281474976710655,
		RefundExpiry:        281474976710655,
		MinFeeBps:           0,
		MaxFeeBps:           100,
		FeeReceiver:         "0x4444444444444444444444444444444444444444",
		Salt:                "0x0000000000000000000000000000000000000000000000000000000000000001",
	}
}

func TestComputePayerAgnosticPaymentInfoHash_Golden84532(t *testing.T) {
	hash, err := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), mockPaymentInfo())
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	want := "0x341988b065a5131b3a82818eb7aba9010135f326af1af7695fce4d2bbebd0b76"
	if hash != want {
		t.Fatalf("hash = %q, want %q", hash, want)
	}
}

func TestComputePayerAgnosticPaymentInfoHash_Golden8453(t *testing.T) {
	hash, err := ComputePayerAgnosticPaymentInfoHash(big.NewInt(8453), mockPaymentInfo())
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	want := "0xa393f8f76a2327a7678488b2d504bda611b7586bb3f334b255a11bb5a75e79ca"
	if hash != want {
		t.Fatalf("hash = %q, want %q", hash, want)
	}
}

func TestComputePayerAgnosticPaymentInfoHash_Golden84532_V1_0(t *testing.T) {
	hash, err := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), mockPaymentInfo(), AuthCaptureEscrowV1_0Address)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	want := "0x19de8ffcb747e5caadb3dda7435cf54992e87cdf0c90e5315ffa129dbb22461e"
	if hash != want {
		t.Fatalf("hash = %q, want %q", hash, want)
	}
}

func TestComputePayerAgnosticPaymentInfoHash_Golden8453_V1_0(t *testing.T) {
	hash, err := ComputePayerAgnosticPaymentInfoHash(big.NewInt(8453), mockPaymentInfo(), AuthCaptureEscrowV1_0Address)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	want := "0x198bbfeaab2f8e36302c662ae41bceaef47a1eb4bf2549cb87aa8daa7f7bb43a"
	if hash != want {
		t.Fatalf("hash = %q, want %q", hash, want)
	}
}

func TestComputePayerAgnosticPaymentInfoHash_Format(t *testing.T) {
	hash, err := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), mockPaymentInfo())
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if len(hash) != 66 || hash[:2] != "0x" {
		t.Fatalf("unexpected hash format: %q", hash)
	}
}

func TestComputePayerAgnosticPaymentInfoHash_Deterministic(t *testing.T) {
	info := mockPaymentInfo()
	a, err := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), info)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	b, err := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), info)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if a != b {
		t.Fatalf("hashes differ: %q vs %q", a, b)
	}
}

func TestComputePayerAgnosticPaymentInfoHash_DifferentChainID(t *testing.T) {
	info := mockPaymentInfo()
	a, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), info)
	b, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(8453), info)
	if a == b {
		t.Fatal("expected different hashes for different chain IDs")
	}
}

func TestComputePayerAgnosticPaymentInfoHash_DifferentPaymentInfo(t *testing.T) {
	info := mockPaymentInfo()
	other := info
	other.MaxAmount = "2000000"
	a, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), info)
	b, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), other)
	if a == b {
		t.Fatal("expected different hashes for different maxAmount")
	}
}

func TestComputePayerAgnosticPaymentInfoHash_DifferentSalt(t *testing.T) {
	info := mockPaymentInfo()
	other := info
	other.Salt = "0x0000000000000000000000000000000000000000000000000000000000000002"
	a, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), info)
	b, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), other)
	if a == b {
		t.Fatal("expected different hashes for different salt")
	}
}

func TestComputePayerAgnosticPaymentInfoHash_PayerAgnostic(t *testing.T) {
	base := mockPaymentInfo()
	a := base
	a.Payer = "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
	b := base
	b.Payer = "0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
	c := base
	c.Payer = "0x0000000000000000000000000000000000000000"

	hashA, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), a)
	hashB, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), b)
	hashC, _ := ComputePayerAgnosticPaymentInfoHash(big.NewInt(84532), c)
	if hashA != hashB || hashA != hashC {
		t.Fatalf("expected payer-agnostic hashes to match: %q %q %q", hashA, hashB, hashC)
	}
}

func TestGenerateSalt(t *testing.T) {
	s1, err := GenerateSalt()
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	s2, err := GenerateSalt()
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if len(s1) != 66 || s1[:2] != "0x" {
		t.Fatalf("bad salt format: %q", s1)
	}
	if s1 == s2 {
		t.Fatal("expected unique salts")
	}
}

func TestSaltBinding(t *testing.T) {
	authorizer := "0x1111111111111111111111111111111111111111"
	zero := "0x0000000000000000000000000000000000000000"
	nonce := "0x0000000000000000000000000000000000000000000000000000000000000abc"

	if IsSaltBindingOn(AuthCaptureExtra{}) {
		t.Fatal("expected binding off for empty extra")
	}
	if IsSaltBindingOn(AuthCaptureExtra{ReceiverAuthorizer: zero, Policy: zero}) {
		t.Fatal("expected binding off for zero addresses")
	}
	if !IsSaltBindingOn(AuthCaptureExtra{ReceiverAuthorizer: authorizer}) {
		t.Fatal("expected binding on for non-zero receiverAuthorizer")
	}

	a, err := DeriveBoundSalt(authorizer, zero, nonce)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	b, err := DeriveBoundSalt(authorizer, zero, nonce)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if a != b {
		t.Fatalf("expected deterministic bound salt, got %q vs %q", a, b)
	}
	want := "0xd0967e09b6c8fccf96277d95a03e98583e8605ab10858b1349aa50ea6d78132c"
	if a != want {
		t.Fatalf("bound salt = %q, want %q", a, want)
	}

	base := a
	changedAuthorizer, _ := DeriveBoundSalt("0x2222222222222222222222222222222222222222", zero, nonce)
	if changedAuthorizer == base {
		t.Fatal("expected bound salt to change when receiverAuthorizer changes")
	}
	changedNonce, _ := DeriveBoundSalt(authorizer, zero, "0x0000000000000000000000000000000000000000000000000000000000000abd")
	if changedNonce == base {
		t.Fatal("expected bound salt to change when saltNonce changes")
	}
}

func TestTypeHashConstants(t *testing.T) {
	if SaltBindingTypeHash.Hex() != "0x8a2a7e41a0bda000ded071ff38b79401d2603e1826516ff2635b11fe9e30877f" {
		t.Fatalf("SaltBindingTypeHash = %s", SaltBindingTypeHash.Hex())
	}
	if PaymentInfoTypeHash.Hex() != "0xae68ac7ce30c86ece8196b61a7c486d8f0061f575037fbd34e7fe4e2820c6591" {
		t.Fatalf("PaymentInfoTypeHash = %s", PaymentInfoTypeHash.Hex())
	}
}
